### PR TITLE
Propagate exceptions from PrefetchingStream

### PR DESCRIPTION
**Pull request by leezu**
## Description ##
This adds logic to PrefetchingStream to report exceptions during setup and prefetching. Based on that, tests for #397 are added. 

### Essentials ###
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage
- [X] Code is well-documented

### Changes ###
- [X] Propagate exceptions from PrefetchingStream workers

## Comments ##
- While checking for any other DataStreams that may suffer from similar issues as in #397 I found an unused file in the scripts folder and deleted it.
- Also wraps the generator  in `_EmbeddingCenterContextBatchify.__iter__`. While an implementation detail (unlike the lazy transformation of streams), it improves consistency throughout the codebase

### CODE REVIEW

File: src/gluonnlp/data/stream.py [hunk 1]
```diff
@@ -221,7 +221,11 @@ def run(self):
         """Method representing the process’s activity."""
         random.seed(self.seed)
         np.random.seed(self.np_seed)
-        mx.random.seed(self.mx_seed)
+        if not isinstance(self, multiprocessing.Process):
+            # Calling mxnet methods in a subprocess will raise an exception if
+            # mxnet is built with GPU support
+            # https://github.com/apache/incubator-mxnet/issues/4659
+            mx.random.seed(self.mx_seed)
```
Comment (leezu) [inline src/gluonnlp/data/stream.py]: @szha @eric-haibin-lin FYI, we cannot guarantee the execution of any mxnet related code would work inside a subprocess.

**Comment by mli** [comment]
Job PR-406/9 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-406/9/index.html

File: src/gluonnlp/data/stream.py [hunk 2]
```diff
@@ -213,57 +221,97 @@ def run(self):
...
+            return next_item
+        else:
+            self._controlq.put(None)
+            if isinstance(next_error[0], StopIteration):
+                raise StopIteration
+            else:
+                return self._reraise(*next_error)
+
+    def _reraise(self, e, tb):
+        print('Reraising exception from Prefetcher')
```
Comment (zhreshold) [inline src/gluonnlp/data/stream.py]: is print appropriate here?

File: src/gluonnlp/data/stream.py [hunk 3]
```diff
@@ -213,57 +221,97 @@ def run(self):
...
+        print(tb)
+        raise e
+
+    def _check_start(self):
+        next_error = self._errorq.get(block=True)
```
Comment (zhreshold) [inline src/gluonnlp/data/stream.py]: if _check_start is called multiple times accidentally, it's going to cause deadlock?

File: src/gluonnlp/data/stream.py (hunk 2 above)
Comment (leezu) [reply src/gluonnlp/data/stream.py]: The idea is to replicate the standard behaviour of Python printing a traceback before crashing due to an exception. So I guess `print` is appropriate, but let me know if you have another idea. We should `print` to `stderr`. I will push a change for that, thanks for catching this

File: src/gluonnlp/data/stream.py (hunk 3 above)
Comment (leezu) [reply src/gluonnlp/data/stream.py]: When calling it twice the behaviour is undefined and may cause deadlock (when reaching the end of the stream). I will add an assertion that it is not called twice. Thanks!

**Comment by leezu** [comment]
Rebased, force-pushed and added a commit to address @zhreshold comments. @zhreshold , once you're happy with the code, can you approve the PR so it can be merged? If you have any other comments, please let me know

**Comment by zhreshold** [comment]
@leezu LGTM now. Thanks!

**Pull request merged**